src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +330,169 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
+            private enum ReturnKind
+            {
+                VoidReturning,
+                IntReturning,
+            }
+            // Does this main method need to forward arguments to the user-provided Main?
+            private enum ParamKind
+            {
+                NoArgs,
+                StringArrayArgs,
+            }
+
+            // Task -> Void
+            // Task<_> -> Int32
+            private static TypeSymbol TranslateReturnType(CSharpCompilation compilation, TypeSymbol returnType)
+            {
+                if (returnType.IsGenericTaskType(compilation)) {
+
+                    return compilation.GetSpecialType(SpecialType.System_Int32);
+                }
+                else
+                {
+                    return compilation.GetSpecialType(SpecialType.System_Void);
+                }
+            }
+
+            private static ReturnKind GetReturnKind(CSharpCompilation compilation, TypeSymbol returnType) {
+                if (returnType.IsGenericTaskType(compilation)) {
+                    return ReturnKind.IntReturning;
+                }
+                else
+                {
+                    return ReturnKind.VoidReturning;
+                }
+            }
+
+            internal AsyncForwardEntryPoint(CSharpCompilation compilation, DiagnosticBag diagnosticBag, NamedTypeSymbol containingType, MethodSymbol userMain): 
+                base(containingType, TranslateReturnType(compilation, userMain.ReturnType)) {
+                Debug.Assert(userMain != null);
+
+                _userMain = userMain;
+                _diagnosticBag = diagnosticBag;
+                _compilation = compilation;
+                _paramKind = userMain.ParameterCount == 0 ? ParamKind.NoArgs : ParamKind.StringArrayArgs;
+                _returnKind = GetReturnKind(compilation, userMain.ReturnType);
+
+                switch (_paramKind)
+                {
+                    case ParamKind.StringArrayArgs:
+                        var stringType = _compilation.GetSpecialType(SpecialType.System_String);
+                        var stringArrayType = ArrayTypeSymbol.CreateCSharpArray(_compilation.Assembly, stringType);
+                        _parameters = ImmutableArray.Create(
+                            SynthesizedParameterSymbol.Create(this, stringArrayType, 0, RefKind.None));
+                        break;
+                    default:
+                        _parameters = ImmutableArray<ParameterSymbol>.Empty;
+                        break;
+                }
+            }
+
+            public override string Name => MainName;
+
+            public override ImmutableArray<ParameterSymbol> Parameters => _parameters;
+
+            internal override BoundBlock CreateBody()
+            {
+                var syntax = this.GetSyntax();
+
+                var refKinds = Parameters.Length == 0  ? ImmutableArray<RefKind>.Empty : ImmutableArray.Create(RefKind.None);
```

Comment:
Better yet, `argumentRefKindsOpt` is an optional parameter so we can pass in `default(ImmutableArray<RefKind>)` instead.